Risteys â€“ Web portal
====================

This directory contains the code to power the online FinRegistry-FinnGen Risteys web portal. For the data pipeline, please check the link:../pipeline/README.md[pipeline] directory.

The project is based on the https://phoenixframework.org/[Phoenix web framework].
This framework is written in https://elixir-lang.org/[elixir], which is also used to do data pre-processing and import data into the database.
The front-end interactions are using https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html[Phoenix LiveView].


Setting up a development environment
------------------------------------

Requirements
~~~~~~~~~~~~

- https://elixir-lang.org/install.html[elixir]
- PostgreSQL database


Configuration
~~~~~~~~~~~~~

Some configuration files contain secret credentials, therefore are not present in this public GitHub repository. Risteys developers can get them by requesting access to the GitLab repository `risteys/risteys_secrets`.


Running
~~~~~~~
Make sure you are in the directory `risteys_elixir`, then run the command

[source,bash]
----
mix phx.server
----

You can now access Risteys at http://localhost:4000


Importing data
--------------

Once the Risteys web server is running we still need to put data there.

If you need to compute the statistics by yourself, then check the link:../pipeline/README.md[pipeline documentation].

When you have the data, you can proceed to import it into the Risteys database.
The scripts to import data files into the Risteys database are in this current directory (`risteys_elixir`) and they should be run from that directory, for example:

[source,bash]
----
mix run import_icd9.exs <path-to-input-file>
----


Import scripts should be run in this order:

1. import_icd10.exs, using:
  - <<file-medcode,FG Medcode>>

2. import_icd9.exs, using:
  - <<file-medcode,FG Medcode>>

3. import_endpoint_csv.exs, using:
  - <<file-endp-defs,FG Endpoint definitions>>
  - <<file-endp-main-tag,FG Endpoint main tag>>
  - <<file-endp-taglist,FG Endpoint tag list>>
  - <<file-icd10,ICD-10>>
  - <<file-endp-selected-core,FG Endpoint selected core>>

4. import_intermediate_counts.exs, using:
  - <<file-interm-counts,FG intermediate counts>> and `FG` as dataset argument.

5. import_ontology.exs, using:
  - <<file-ontology,Risteys ontology>>
  - <<file-corrected-descriptions,Risteys corrected endpoint descriptions>>

6. import_excluded_endpoints.exs, using:
  - <<file-excluded_endp,FinRegistry excluded endpoints>>

7. import_key_figures.exs, using:
  - <<file-key-fig-FR-all,Risteys FinRegistry key figures of full population>> and `FR` as dataset argument

8. import_key_figures.exs, using:
  - <<file-key-fig-FR-index,Risteys FinRegistry key figures of index-persons>> and `FR_index` as dataset argument

9. import_key_figures.exs, using:
  - <<file-key-fig-FG,Risteys FinnGen key figures>> and `FG` as dataset argument

10. import_distributions.exs, using:
  - <<file-distrib-age-FR,Risteys FinRegistry age distributions>> and `age` and `FR` as arguments

11. import_distributions.exs, using:
  - <<file-distrib-year-FR,Risteys FinRegistry year distributions>> and `year` and `FR` as arguments

12. import_distributions.exs, using:
  - <<file-distrib-age-FG,Risteys FinnGen age distributions>> and `age` and `FG` as arguments

13. import_distributions.exs, using:
  - <<file-distrib-year-FG,Risteys Finngen year distributions>> and `year` and `FG` as arguments

14. import_stats_cumulative_incidence.exs, using:
  - <<file-cumul-inc-FR,Risteys FinRegistry cumulative incidence>> and `FR` as argument

15. import_stats_cumulative_incidence.exs, using:
  - <<file-cumul-inc-FG,Risteys FinnGen cumulative incidence>> and `FG` as argument

16. import_interactive_mortality_baseline.exs, using:
  - <<file-mortality-baseline,Risteys FinRegistry mortality baseline cumulative hazards>>

17. import_interactive_mortality_params.exs, using:
  - <<file-mortality-params,Risteys FinRegistry mortality parameters>>

18. import_mortality_counts.exs, using:
  - <<file-mortality-counts,Risteys FinRegistry mortality counts>>

19. import_correlation.exs, using:
  - <<file-corr-pheno-geno,FinnGen phenotypic + genotypic correlations>>
  - <<file-coloc-variants,FinnGen coloc variants>>

20. import_case_overlaps_fr.exs, using:
  - <<file-case-overlaps-FR, FinRegistry case overlaps>>

21. import_coxhr.exs, using
  - <<file-surv-FR, FinRegistry survival analysis results>>

22. import_genetic_correlations.exs, using
  - <<file-gen-correlations, FinnGen genetic correlations>>

23. import_genes.exs, using
  - <<file-havana-genes, HAVANA gene list>>

24. import_upset_plots.exs, using
  - <<file-upset-plots, Upset Plots>> and `FG` as dataset argument

25. import_codewas.exs, using
  - <<file-codewas-endpoints, CodeWAS Endpoints>>
  - <<file-codewas-codelist, CodeWAS code list>>


File list
~~~~~~~~~

* [[file-icd10]]ICD-10
** name: `ICD10_koodistopalvelu_2015-08_26_utf8__XXH64_71956a051f960e51.csv`
** source: https://koodistopalvelu.kanta.fi/codeserver/pages/classification-view-page.xhtml?classificationKey=23&versionKey=58[Kela Kansallinen koodistopalvelu]

* [[file-medcode]]FinnGen medcode
** name: `finngen_R9_medcode_ref__XXH64_708053b379a04020.tsv`
** source: FinnGen - Library Green

* [[file-endp-defs]]FinnGen endpoint definitions
** name: `finngen_R12_endpoint_core_noncore_1.0.added_omit2__XXH64_399efaa48ca282b6.csv`
** source: Merging of `finngen_R12_endpoint_core_noncore_1.0.xlsx` and `OMIT` column from `Endpoints_Controls_FINNGEN_ENDPOINTS_DF12_Final_2023-05-17.xlsx` - FinnGen clinical team - https://github.com/FINNGEN/Df12-endpoint-and-control-definitions/[GitHub]

* [[file-endp-main-tag]]FinnGen endpoint main tag
** name: `FINNGEN_ENDPOINTS_DF12_Final_2023-05-17.names_tagged_ordered__XXH64_8264f1235f3f7221.txt`
** source: FinnGen clinical team - https://github.com/FINNGEN/Df12-endpoint-and-control-definitions/[GitHub]

* [[file-endp-taglist]]FinnGen endpoint tag list
** name: `TAGLIST_DF12__XXH64_2c6dae042382fea9.csv`
** source: FinnGen clinical team - https://github.com/FINNGEN/Df12-endpoint-and-control-definitions/[GitHub]

* [[file-endp-selected-core]]FinnGen endpoint selected core
** name: `finngen_correlation_clusters_DF8__XXH64_0d9f3a10306791f5.csv`
** source: FinnGen clinical team

* [[file-interm-counts]]FinnGen endpoint intermediate counts
** name: `finngen_endpoints_intermediate_counts_green_export_R12_v1__XXH64_85e199bb39d62337.txt`
** source: FinnGen registry team

* [[file-ontology]]Risteys ontology
** name: `finngen_ontology_2022-08-22__XXH64_2a8d4690fa4ae89a.json`
** source: Risteys pipeline

* [[file-corrected-descriptions]]Risteys corrected endpoint description
** name: `corrected-endpoint-descriptions.airtable-export.2023-10-10__XXH64_b0bd5eb161441ba9.csv`
** source: Risteys Airtable

* [[file-excluded_endp]]FinRegistry excluded endpoints
** name: `excluded_endpoints_FR_Risteys_R12__XXH64_508917188be68559.csv`
** source: Risteys script `exclude_endpoints_finregistry.py`

* [[file-key-fig-FR-all]]Risteys FinRegistry key figures, all individuals
** name: `key_figures_all_2022-10-10_with_EXALLC_EXMORE__XXH64_920b310de04e72e7.csv`
** source: Risteys pipeline

* [[file-key-fig-FR-index]]Risteys FinRegistry key figures, only index-persons
** name: `key_figures_index_2022-10-10_with_EXALLC_EXMORE__XXH64_c62d6a466a0512a1.csv`
** source: Risteys pipeline

* [[file-key-fig-FG]]Risteys FinnGen key figures
** name: `key_figures_all_2023-09-20__XXH64_93a5ba6f09958693.csv`
** source: Risteys pipeline

* [[file-distrib-age-FR]]Risteys FinRegistry age distributions
** name: `distribution_age_2022-10-10_with_EXALLC_EXMORE__XXH64_edd7be5c03a84317.csv`
** source: Risteys pipeline

* [[file-distrib-year-FR]]Risteys FinRegistry year distributions
** name: `distribution_year_2022-10-10_with_EXALLC_EXMORE__XXH64_a5ea390cd797b6e3.csv`
** source: Risteys pipeline

* [[file-distrib-age-FG]]Risteys FinnGen age distributions
** name: `distribution_age_2023-09-20__XXH64_0ab1f53d7d3013f7.csv`
** source: Risteys pipeline

* [[file-distrib-year-FG]]Risteys FinnGen year distributions
** name: `distribution_year_2023-09-20__XXH64_90ee66ed48dfb5fb.csv`
** source: Risteys pipeline

* [[file-cumul-inc-FR]]Risteys FinRegistry cumulative incidence
** name: `cumulative_incidence_2022-10-10_with_EXALLC_EXMORE__XXH64_c08ae173edf55e72.csv`
** source: Risteys pipeline

* [[file-cumul-inc-FG]]Risteys FinnGen cumulative incidence
** name: `all_cumulative_incidence__r12__2023-09-20__XXH64_f4909d1f5b2565ee.csv`
** source: Risteys pipeline

* [[file-mortality-baseline]]Risteys FinRegistry mortality baseline cumulative hazards
** name: `mortality_baseline_cumulative_hazard_2022-10-11_with_EXALLC_EXMORE__XXH64_0088608aa7e021bd.csv`
** source: Risteys pipeline

* [[file-mortality-params]]Risteys FinRegistry mortality parameters
** name: `mortality_params_2022-10-11_with_EXALLC_EXMORE__XXH64_8f4fdc15e1c061c1.csv`
** source: Risteys pipeline

* [[file-mortality-counts]]Risteys FinRegistry mortality counts
** name: `mortality_counts_2022-10-11_with_EXALLC_EXMORE__XXH64_f7f9581772ec80c6.csv`
** source: Risteys pipeline

* [[file-corr-pheno-geno]]FinnGen phenotypic + genotypic correlations
** name: `corr_pheno-fg-r12.0_geno-fg-r12.0_full-join__2023-11-15__XXH64_86c9c7a833d663e5.csv.zst`
** source: https://github.com/FINNGEN/endpcorr[FinnGen correlation pipeline] for the phenotypic file, merged with genotypic correlation file from FinnGen analysis team

* [[file-coloc-variants]]FinnGen coloc variants
** name: `r12.autoreport.compare.keep_cs.r2_0.8.pval_5e_8.variants__XXH64_854d4aeb62e9664a.csv`
** source: FinnGen analysis team

* [[file-case-overlaps-FR]] FinRegistry case overlaps
** name: `case_overlap_2022-12-31__XXH64_fb1ca5ba80e4a0ba.csv.zst`
** source: Risteys pipeline

* [[file-surv-FR]] FinRegistry survival analysis results
** name: `surv_priority_endpoints_2022-12-25__XXH64_b92220411f705ef2.csv`
** source: Risteys pipeline

* [[file-gen-correlations]] FinnGen genetic correlations
** name: `finngen_R12_FIN.ldsc.summary__XXH64_40dc9830272f8976.tsv`
** source: FinnGen Green library

* [[file-havana-genes]] HAVANA gene list
** name: `havana__XXH64_085a38684d85191e.json`
** source: HAVANA through FinnGen

* [[file-upset-plots]], Upset Plots
** name: `upset_plots_R12__censor_below_5__no_finngenids__2023-10-24__XXH64_84ac03a5853d9be4.tar.zst`
** source: Harri S, FinnGen Phenotype team

* [[file-codewas-endpoints]] CodeWAS Endpoints
** name: `codewas_endpoints_r11.filtered_nlog10p.green.2023-10-31.XXH64_3099f7b7f82bd251.jsonl.zst`
** source: Raw CodeWAS data from Javier G-T, FinnGen Phenotype Team; then applied link:../pipeline/risteys_pipeline/finngen/filter_codewas_greendata.py[`filter_codewas_greendata.py`]

* [[file-codewas-codelist]] CodeWAS code list
** name: `medical_codes_fg_code_info_v3_fg_codes_info_v3.csv`
** source: Javier G-T, FinnGen Phenotype Team

* [[file-omop-loinc-concept]] OMOP LOINC â€“ Concepts
** name: `CONCEPT.xsv_fmt.XXH64_34860affdef58dc2.csv`
** source: Downloaded LOINC `CONCEPT.csv` from https://athena.ohdsi.org/vocabulary/list[OHDSI Athena], CODE (CDM V5) = LOINC, Latest update = 18-Sep-2023, then applied proper CSV formatting (with `xsv fmt -d"\t"`) to fix bad escaping in original `CONCEPT.csv`.

* [[file-omop-loinc-concept-relationship]] OMOP LOINC â€“ Concept relationships
** name: `CONCEPT_RELATIONSHIP.xsv_fmt.XXH64_e04d06ecdd63e0c1.csv`
** source: Downloaded LOINC `CONCEPT.csv` from https://athena.ohdsi.org/vocabulary/list[OHDSI Athena], CODE (CDM V5) = LOINC, Latest update = 18-Sep-2023, then applied proper CSV formatting (with `xsv fmt -d"\t"`) to fix bad escaping in original `CONCEPT_RELATIONSHIP.csv`.

* [[file-kanta-lab-omopid-list]] Kanta lab â€“ List of OMOP Concept IDs
** name: `all_omop_ids.XXH64_6ae9ea63adfb491a.jsonl`
** source: Risteys Kanta lab pipeline
